### ISSUE TITLE
SUModelCreate always returns unknown exception for intel Macs running SU2023

### ISSUE THREAD

**Issue by Whaat** [SketchUp/api-issue-tracker#891]
SketchUp Pro 23.0.396
Intel Mac OSX 12.4

I have had multiple support requests about this issue and in each case, the issue occurs when the user is running SU2023 and an intel Mac.  It does not occur for arm64 Macs or for older versions of SketchUp.

Our Skimp plugin does not work for SU2023 intel Macs because of this issue.  I do not think there is any workaround.

```c
	SUInitialize();
	SUModelRef model = SU_INVALID;
	SUResult res = SUModelCreate(&model);
	if (res != SU_ERROR_NONE)
		// res will be equal to SU_ERROR_UNKNOWN_EXCEPTION so this block will execute
```
The live version of Skimp is using the SketchUp API Framework universal binary version that was released for SU2022.  In my current dev version of Skimp, I updated the API Framework to the current version (SU2023) thinking it might solve the problem.  But the issue remains.

Download test code here:
[skimp-temp-sdk-bugreport.zip](https://github.com/SketchUp/api-issue-tracker/files/11251077/skimp-temp-sdk-bugreport.zip)

The test code is a modified and heavily purged branch of the Skimp repo.

Instructions to test:

- Ensure you are using SU2023 and an intel Mac
- Open the xcode project at c++/MSS_Licensing/MSS_Licensing.xcodeproj
- Ensure DM_skimp.rb will load when SU2023 runs. (I have my own setup for this but you will need to do your own method that will work for your environment)
- Run the active scheme by clicking the 'play' button in Xcode.  It should build the Ruby 2.7 target (MSS_Licensing.bundle) and launch SU2023 (if you have it located in /Applications)
- If it builds and loads correctly, DM_skimp.rb will call a function from the C extension called 'testSUModelCreate' which should trigger the bug

**Comment by Whaat** [comment]
I would say it is using the C API within SketchUp (triggering C API calls from a Ruby extension) but not the Live API.  I can reproduce it every time myself, not just from end users.

**Comment by thomthom** [comment]
> I would say it is using the C API within SketchUp (triggering C API calls from a Ruby extension) but not the Live API.

So this is acting on a separate model from within a Ruby extension, using the C API? And are you linking against SketchUpAPI.framework or the application?

> Have you been able to reproduce? If not, I can send you a simple test xcode project that reproduces the issue.

Please do, this sounds somewhat specific of a scenario. I'd rather not guess at the setup here.

**Comment by Whaat** [comment]
I have modified the first post and included test code with instructions.  I am happy to do an online meeting to walk you through the issue if you have trouble reproducing. Thanks.

**Comment by DanRathbun** [comment]
I'm not sure if the bot can access dropbox. You can just attach the zip archive to the post proper.

**Comment by GWydouw** [comment]
This sounds like a similar problem we have with adding SketchUp models to layout with the ruby api.

**Comment by DanRathbun** [comment]
Guy, please open an issue for the LayOut Ruby API if one does not yet exist.

**Comment by Whaat** [comment]
@thomthom Do you have everything you need?  Any idea when this will be looked at?  It is a critical bug that prevents Skimp from working on Intel Macs.  I have customers asking me when it might be fixed.  Thanks

**Comment by thomthom** [comment]
For the record, this was crashing because it was linking against the standalone C API while being used within SketchUp. While using the C API "live" it needs to be setup such that it links against the SketchUp application;

https://github.com/SketchUp/sketchup-live-c-api/blob/main/README.md#macos

**Comment by Whaat** [comment]
Thanks Thom - this seems to do the trick.  I suggest some changes be made to the SDK documentation as it is not clear that developers MUST link to the 'live' API if calling it from inside a Ruby C Extension.  The documentation only says that you need to link to it if you plan to read from the open model (which I'm not in my unique case)
https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/index.html

Even under the 'Extension Usage' section of the docs above, it says that it is 'possible to use the C API for read only access to the open model via...'.  There is no indication that linking to the Live C API is needed for any other purpose.  This needs to be cleared up in the docs...especially because it might still appear to work 100% fine (like in my situation for arm64 Macs and Windows all versions).

I never ran into any issues like this for years and I've always been linking to SketchUpAPI.lib (not sketchup.lib or the Mac equivalent).  It would be good if SketchUp could detect whether an extension is not linking correctly and raise an error as they sorts of issues are really tough to diagnose and fix, at least for me.

**Comment by DanRathbun** [comment]
> _It would be good if SketchUp could detect whether an extension is not linking correctly and raise an error as they sorts of issues are really tough to diagnose and fix, at least for me._

Could this be done with an IDE plugin? Sort of a "project cop" for Objective C in Xcode?

(I have no experience with Mac coding myself, just musing.)

**Comment by DanRathbun** [comment]
> _I suggest some changes be made to the SDK documentation as it is not clear that developers MUST link to the 'live' API if calling it from inside a Ruby C Extension. The documentation only says that you need to link to it if you plan to read from the open model (which I'm not in my unique case) ..._

**I agree**. In fact, the [C API home page](https://extensions.sketchup.com/developers/sketchup_c_api/sketchup/index.html) actually has a bullet item under **Standalone Usage** typical cases that says ...
> * _**SketchUp extension** reading/writing .skp files **separate from the open model**._

... and then directly below telling coders to link against `SketchUpAPI.lib` and `SketchUpAPI.framework`.

This and the [README for the Live API example repository](https://github.com/SketchUp/sketchup-live-c-api/blob/main/README.md) are in direct conflict.

**Comment by DanRathbun** [comment]
> _I never ran into any issues like this for years and I've always been linking to SketchUpAPI.lib (not sketchup.lib or the Mac equivalent)._

I need to poke you in fun here Dale, and [show you where **3 years ago** Thomas told you to link correctly](https://forums.sketchup.com/t/best-practice-to-support-multiple-sketchup-versions/96018/7) ...
> _Are you linking against sketchup.lib or SketchUpAPI.lib? The former links to the C API within Sketchup.exe, the latter links against the standalone version. Mixing those two doesn’t work well._

But again, I also agree that the C API home page appears to lead programmers astray.

**Comment by Whaat** [comment]
Wow Dan - can't believe you dug that up lol.
I guess I was warned after all.  Sometimes, using the "if it ain't broke, don't fix it" approach in coding can bite you one day

**Comment by DanRathbun** [comment]
Hehe ... yeah, I have a good memory for details. ; )

**Comment by Whaat** [comment]
I have a new twist to report.  Perhaps it warrants a separate issue being opened. 

Based on this discussion above, I updated Skimp to link to sketchup.lib instead of SketchUpAPI.lib on Windows builds.

Now, after releasing a new version with this change, I have received a report of crashes where previously everything worked fine.  I did some testing and, sure enough, the crash only occurs when linking to sketchup.lib.  No crashes when linking to SketchUpAPI.lib.

So now, to quickly patch my crashing live version of Skimp, I need to release an update that links to SketchUpAPI.lib even though this goes against the recommendations provided.

I'm not sure how to submit this bug report as it could be hard to reproduce without providing access to the entire Skimp repo.  Would submitting the Windows crash reports help?
